### PR TITLE
Service desk plus on premise

### PR DESCRIPTION
**Pull request by lm-t**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
ServiceDeskPlus On-Premises version.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [x] 5.0.0
- [ ] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by lm-t** [comment]
@guykeller I've submitted the PR. I'm waiting on my team to details on our 'partner ID' so that I can finish completing the Contributor Details Questionnaire.

**Review by guykeller — CHANGES_REQUESTED** [review]

File: Packs/ServiceDeskPlus_On_Premise/Integrations/ServiceDeskPlus_On_Premise/README.md [hunk 1]
```diff
@@ -0,0 +1,736 @@
...
+| closure_comments | The comments that should be added when closing the request | Optional |
+| closure_code | The closure code to add to the request. Can be "Cancelled", "Failed", "Postponed", "Rejected", "Success" or "Unable to Reproduce" | Optional |
+| requester_ack_comments | The requester comments that should be added to the request | Optional |
+| requester_ack_resolution | Boolean. | Optional |
+| request_id | The id of the request that should be closed | Required |
+
+
+#### Context Output
+
+There is no context output for this command.
```
[the remark this answers is not in the document]
Comment (lm-t) [reply Packs/ServiceDeskPlus_On_Premise/Integrations/ServiceDeskPlus_On_Premise/README.md:724]: The context outputs were left as is from the original integration.

File: Packs/ServiceDeskPlus_On_Premise/Integrations/ServiceDeskPlus_On_Premise/ServiceDeskPlus_On_Premise_description.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+1. Enter the Server URL info.
+2. Enter the Technician Key.
+3. Click the **Test** button to validate the instance.
+
+**NOTES**
+- For more details about generating a technician key please refer to the [help documentation](https://help.servicedeskplus.com/api/rest-api.html$key)
+
+## Fetch-Incidents Query
+
+Filters should be in the format "{field':<field_name>, 'condition':<condition>, 'values':'val_1,val_2', 'logical_operator':<op>}".
```
[the remark this answers is not in the document]
Comment (lm-t) [reply Packs/ServiceDeskPlus_On_Premise/Integrations/ServiceDeskPlus_On_Premise/ServiceDeskPlus_On_Premise_description.md:15]: The parameter's description has a reference to "see (?)" section. This was left as is from the original integration.

File: Packs/ServiceDeskPlus_On_Premise/pack_metadata.json [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    "author": "Cortex XSOAR",
+    "url": "https://www.paloaltonetworks.com/cortex",
+    "email": "",
+    "created": "2021-01-21T17:06:14Z",
+    "categories": [
+        "Case Management"
+    ],
+    "tags": [],
+    "useCases": [],
+    "keywords": []
```
[the remark this answers is not in the document]
Comment (lm-t) [reply Packs/ServiceDeskPlus_On_Premise/pack_metadata.json:15]: The original integration does not have use cases, tags, nor keywords.

File: Packs/ServiceDeskPlus_On_Premise/Integrations/ServiceDeskPlus_On_Premise/ServiceDeskPlus_On_Premise.py [hunk 4]
```diff
@@ -0,0 +1,782 @@
...
+                try:
+                    return res.json()
+                except ValueError as exception:
+                    raise DemistoException('Failed to parse json object from response: {}'
+                                           .format(res.content), exception)
+
+            if res.status_code in [401]:
+                if demisto.getIntegrationContext().get('expiry_time', 0) <= date_to_timestamp(datetime.now()):
+                    # self.get_access_token()
+                    return self.http_request(method, url_suffix, full_url=full_url, params=params)
```
[the remark this answers is not in the document]
Comment (lm-t) [reply Packs/ServiceDeskPlus_On_Premise/Integrations/ServiceDeskPlus_On_Premise/ServiceDeskPlus_On_Premise.py]: Removed the logic. It was from the original integration and it is not needed.

File: Tests/conf.json [hunk 5]
```diff
@@ -3182,6 +3182,10 @@
             "integrations": "FeedExpanse",
             "playbookID": "Feed Expanse Test",
             "fromversion": "6.0.0"
+        },
+        {
+            "integrations": "ServiceDeskPlus (On-Premise)",
+            "playbookID": "<TestPlaybook ID>"
```
[the remark this answers is not in the document]
Comment (lm-t) [reply Tests/conf.json]: Added the test playbook ID.

**Review by guykeller — APPROVED** [review]

File: Packs/ServiceDeskPlus_On_Premise/Integrations/ServiceDeskPlus_On_Premise/ServiceDeskPlus_On_Premise.yml [hunk 6]
```diff
@@ -0,0 +1,801 @@
...
+    - contextPath: ServiceDeskPlus.Request.CancellationRequested
+      description: Indicates whether a cancellation was requested.
+      type: Boolean
+    - contextPath: ServiceDeskPlus.Request.HasNotes
+      description: Indicates whether the command has notes.
+      type: Boolean
+  - arguments:
+    - default: false
+      description: The ID of the request to delete.
+      isArray: false
```
Comment (guykeller) [inline Packs/ServiceDeskPlus_On_Premise/Integrations/ServiceDeskPlus_On_Premise/ServiceDeskPlus_On_Premise.yml:143]: If it accepts a comma-separated list, this should be true

File: Packs/ServiceDeskPlus_On_Premise/Integrations/ServiceDeskPlus_On_Premise/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (guykeller) [reply Packs/ServiceDeskPlus_On_Premise/Integrations/ServiceDeskPlus_On_Premise/README.md:724]: Same here - doesn't mean this new integration shouldn't have them.
What do the commands with no outputs return? Do we really think they should remain as such?
Could you please send me the API docs and refer me to the proper place?

**Review by guykeller — CHANGES_REQUESTED** [review]

File: Packs/ServiceDeskPlus_On_Premise/Integrations/ServiceDeskPlus_On_Premise/ServiceDeskPlus_On_Premise.yml (hunk 6 above)
Comment (lm-t) [reply Packs/ServiceDeskPlus_On_Premise/Integrations/ServiceDeskPlus_On_Premise/ServiceDeskPlus_On_Premise.yml]: Set isArray to true after verifying.

File: Packs/ServiceDeskPlus_On_Premise/pack_metadata.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (lm-t) [reply Packs/ServiceDeskPlus_On_Premise/pack_metadata.json]: Added pack metadata.

File: Packs/ServiceDeskPlus_On_Premise/Integrations/ServiceDeskPlus_On_Premise/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (lm-t) [reply Packs/ServiceDeskPlus_On_Premise/Integrations/ServiceDeskPlus_On_Premise/README.md:724]: I looked at the commands that do not have context outputs and they do not produce any context outputs. The commands either DELETE or PUT changes into the API endpoint thus not receiving any outputs to be displayed in Demisto.

[the remark this answers is not in the document]
Comment (lm-t) [reply Packs/ServiceDeskPlus_On_Premise/Integrations/ServiceDeskPlus_On_Premise/README.md:724]: API Docs -> [here](https://ui.servicedeskplus.com/APIDocs3/index.html)

**Comment by guykeller** [comment]
@lm-t 
Can you please do the followiing:
1. Invite me as a contributor to your branch, as I seem to be unable to make changes
2. Pull from Master on your base branch, and from the base branch on your contribution branch
Thanks!

**Review by guykeller — APPROVED** [review]

**Pull request merged**